### PR TITLE
hostapd: 2.9 -> 2.10; wpa_supplicant: 2.9 -> 2.10

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change
https://w1.fi/security/2022-1/sae-eap-pwd-side-channel-attack-update-2.txt
https://w1.fi/cgit/hostap/plain/hostapd/ChangeLog
https://w1.fi/cgit/hostap/plain/wpa_supplicant/ChangeLog

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/wpa_supplicant/default.nix [hunk 1]
```diff
@@ -8,45 +8,16 @@
...
+    sha256 = "sha256-IN965RVLODA1X4q0JpEjqHr/3qWf50/pKSqR0Nfhey8=";
   };
 
   patches = [
-    (fetchurl {
-      name = "CVE-2019-16275.patch";
-      url = "https://w1.fi/security/2019-7/0001-AP-Silently-ignore-management-frame-from-unexpected-.patch";
-      sha256 = "15xjyy7crb557wxpx898b5lnyblxghlij0xby5lmj9hpwwss34dz";
-    })
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/wpa_supplicant/default.nix:26]: Please remove the unused fetchpatch from inputs.

File: pkgs/os-specific/linux/hostapd/default.nix [hunk 2]
```diff
@@ -16,38 +16,8 @@ stdenv.mkDerivation rec {
...
-      url = "https://w1.fi/security/2020-1/0002-WPS-UPnP-Fix-event-message-generation-using-a-long-U.patch";
-      sha256 = "1pk08b06b24is50bis3rr56xjd3b5kxdcdk8bx39n9vna9db7zj9";
-    })
-    (fetchurl {
-      name = "CVE-2020-12695_0003-WPS-UPnP-Handle-HTTP-initiation-failures-for-events-.patch";
-      url = "https://w1.fi/security/2020-1/0003-WPS-UPnP-Handle-HTTP-initiation-failures-for-events-.patch";
-      sha256 = "12npqp2skgrj934wwkqicgqksma0fxz09di29n1b5fm5i4njl8d8";
-    })
-    # In wpa_supplicant and hostapd 2.9, forging attacks may occur because AlgorithmIdentifier parameters are mishandled in tls/pkcs1.c and tls/x509v3.c.
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/hostapd/default.nix:47]: Please remove the unused fetchpatch from inputs.

File: pkgs/os-specific/linux/hostapd/default.nix [hunk 3]
```diff
@@ -16,38 +16,8 @@ stdenv.mkDerivation rec {
...
-      name = "CVE-2021-30004.patch";
-      url = "https://w1.fi/cgit/hostap/patch/?id=a0541334a6394f8237a4393b7372693cd7e96f15";
-      sha256 = "1gbhlz41x1ar1hppnb76pqxj6vimiypy7c4kq6h658637s4am3xg";
+      url = "https://raw.githubusercontent.com/openwrt/openwrt/eefed841b05c3cd4c65a78b50ce0934d879e6acf/package/network/services/hostapd/patches/300-noscan.patch";
```
Comment (NinjaTrappeur) [inline pkgs/os-specific/linux/hostapd/default.nix:19]: Is there a particular reason for this URL change? The pointed content seems identical.

Comment (NinjaTrappeur) [reply pkgs/os-specific/linux/hostapd/default.nix:19]: I guess explicitly pointing to a specific commit pins a bit further the content, sounds good to me.

**Review by picnoir — APPROVED** [review]
Manually tested this on a couple of deployments. 

LGTM

File: pkgs/os-specific/linux/hostapd/default.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/os-specific/linux/hostapd/default.nix:19]: Yeah, it was scratching an itch.

File: pkgs/os-specific/linux/hostapd/default.nix [hunk 4]
```diff
@@ -16,38 +16,8 @@ stdenv.mkDerivation rec {
     (fetchurl {
       # Note: fetchurl seems to be unhappy with openwrt git
       # server's URLs containing semicolons. Using the github mirror instead.
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/hostapd/default.nix:18]: Nit: Since we are not using the github link we could have also used fetchpatch.

**Comment by vcunat** [comment]
For me it shows the window well, but my machine does not have wpa_supplicant or WiFi.

**Comment by ncfavier** [comment]
Ok so the problem is that my list of networks (`wpa_cli list_networks`) looks like

```
0   mutable0
1   mutable1
2   mutable2
...
0   immutable0
1   immutable1
```

(I am using `networking.wireless.allowAuxiliaryImperativeNetworks`)

and the logic in that commit assumes that the IDs are strictly increasing, otherwise it might get stuck in an infinite loop.

The fact that the immutable network IDs start again from 0 is probably a bug in our `0001-Implement-read-only-mode-for-ssids.patch`, cc @Ma27

**Comment by Ma27** [comment]
> Definitely an upstream bug.

Well, we (to be precise, I) messed around in their code, so not sure if that still counts as upstream bug :) 

I'll take a look.

**Comment by ncfavier** [comment]
I mean that it would still happen without your patch.

**Comment by ncfavier** [comment]
Using the patch suggested above solves the issue:

```
--- a/wpa_supplicant/config_file.c
+++ b/wpa_supplicant/config_file.c
@@ -297,8 +297,8 @@ struct wpa_config * wpa_config_read(const char *name, struct wpa_config *cfgp)
 	struct wpa_ssid *ssid, *tail, *head;
 	struct wpa_cred *cred, *cred_tail, *cred_head;
 	struct wpa_config *config;
-	int id = 0;
-	int cred_id = 0;
+	static int id = 0;
+	static int cred_id = 0;

 	if (name == NULL)
 		return NULL;
```

**Comment by Ma27** [comment]
Seems reasonable to me. Do you want to file a PR? %)

**Comment by ncfavier** [comment]
Yes, I'll send the patch to their mailing list.

**Comment by Ma27** [comment]
I'd say it's OK to file a PR here as well :)

**Comment by ncfavier** [comment]
Oh, yes, will do that as well.

**Comment by ncfavier** [comment]
Opened https://github.com/NixOS/nixpkgs/pull/158505 and sent the patch upstream

**Pull request merged**